### ISSUE TITLE
Async CoreData stack initialization when using iCloud

### ISSUE THREAD

**Issue by pronebird** [magicalpanda/MagicalRecord#1054]
Hi,

From time to time when I start my app I get a blank table view controller because FRC reports there are no objects. Although if I force refresh from interface and re-create FRC then data shows up.

I have a very simple configuration for iCloud:

```objc
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [MagicalRecord setLoggingLevel:MagicalRecordLoggingLevelVerbose];
    [MagicalRecord setupCoreDataStackWithiCloudContainer:nil contentNameKey:nil localStoreNamed:@"CoreData.sqlite" cloudStorePathComponent:nil completion:^{
        [self _addSeedData]; // added only once on first load
    }];
}
```

Log:

```
Created new private queue context: <NSManagedObjectContext: 0x7c46a860>
Created new context Untitled Context with store coordinator: <NSPersistentStoreCoordinator: 0x7bf845d0>
Set root saving context: <NSManagedObjectContext: 0x7c46a860>
Created new main queue context: <NSManagedObjectContext: 0x7beb0c60>
Set default context: <NSManagedObjectContext: 0x7beb0c60>
Timeline: viewDidLoad, fetchedObjects.count = 0
iCloud is not enabled
Created new private queue context: <NSManagedObjectContext: 0x7c1bc570>
Created new private queue context: <NSManagedObjectContext: 0x7c1adf40>
Created new private queue context: <NSManagedObjectContext: 0x7c1927c0>
NO CHANGES IN ** saveWithBlockAndWait: ** CONTEXT - NOT SAVING
NO CHANGES IN ** saveWithBlockAndWait: ** CONTEXT - NOT SAVING
NO CHANGES IN ** Untitled Context ** CONTEXT - NOT SAVING
```

FRC setup:

```objc
- (void)viewDidLoad {
    [super viewDidLoad];
    NSFetchedResultsController *controller = [ExpenseModel MR_fetchAllSortedBy:@"placedAt" ascending:NO withPredicate:nil groupBy:@"monthYearSection" delegate:self];
    self.fetchController = controller;
}
```

Looking at this:

https://github.com/magicalpanda/MagicalRecord/blob/c7b14f658b4fca32f8d33f8f76160203053ce4b9/MagicalRecord/Categories/NSPersistentStoreCoordinator%2BMagicalRecord.m#L147

So the whole iCloud setup is asynchronous that's why my root VC does not get any data when it creates FRC, because there is simply no storage added to CD stack at that point yet.

So if I comment dispatch_async and make the whole iCloud setup synchronously then everything works. I can assume that async queue was used to avoid any kind of locking of the main thread but reading objc.io (http://www.objc.io/issues/10-syncing-data/icloud-core-data/#asynchronous-persistent-store-setup) it seems that iCloud does not work that way anymore since iOS 7.

> In iOS 7, calling addPersistentStoreWithType:configuration:URL:options:error: with iCloud options now returns a store almost immediately.
>
> In previous OS versions, this method wouldn’t return until iCloud data was downloaded and merged into the persistent store. This would cause significant delays, meaning that any calls to the method would need to be dispatched to a background queue. Thankfully, this is no longer necessary.

Most likely this wouldn't work on iOS 6.1 which is still a deployment target for MagicalRecord.

I would appreciate any suggestion or workaround to this problem.

**Comment by pronebird** [comment]
According to logs FRC reports 0 objects sometimes. It seems that it takes some time to setup CD stack. How do you deal with this because apparently we have asynchronous setup of CD stack, as opposed to atomic setup when we do not use iCloud

**Comment by pronebird** [comment]
Looking at this:

https://github.com/magicalpanda/MagicalRecord/blob/c7b14f658b4fca32f8d33f8f76160203053ce4b9/MagicalRecord/Categories/NSPersistentStoreCoordinator%2BMagicalRecord.m#L147

So the whole iCloud setup is asynchronous that's why my root VC does not get any data when it creates FRC, because there is simply no storage added to CD stack yet.

This does not seem right.

**Comment by tonyarnold** [comment]
> it seems that iCloud does not work that way anymore since iOS 7

This is great! I don't use Core Data with iCloud in any of my apps, so this bit of the code doesn't get much attention. I'll need to write some code to handle the pre-iOS 7.0 approach, but that's OK — thanks for the clarification!

**Comment by pronebird** [comment]
If you have any suggestion for patch, then I could do that.

**Comment by lassebunk** [comment]
@pronebird I think the simplest solution would be doing an `AndWait` version of the iCloud setup methods.

For the time being, here is a workaround that works with the current version of MagicalRecord:

```
var magicalRecordSetupFinished = false
MagicalRecord.setupCoreDataStackWithiCloudContainer("id", contentNameKey: "key", localStoreNamed: "name", cloudStorePathComponent: nil) {
    magicalRecordSetupFinished = true
}
while !magicalRecordSetupFinished {
    NSRunLoop.currentRunLoop().runMode(NSDefaultRunLoopMode, beforeDate: NSDate.distantFuture())
}
```
